### PR TITLE
[NU-1061] Scenario parameters api

### PR DESCRIPTION
**Pull request by arkadius**
## Checklist before merge
- [ ] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [ ] Code is cleaned from temporary changes and commented out lines
- [ ] Parts of the code that are not easy to understand are documented in the code
- [ ] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [ ] Verify that PR will be squashed during merge

### CODE REVIEW

**Review by lukasz-bigorajski — CHANGES_REQUESTED** [review]

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ScenarioParametersEndpoints.scala [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+object ScenarioParametersEndpoints {
+
+  object Dtos {
+
+    @derive(encoder, decoder, schema)
+    case class ScenarioParametersCombination(
+        processingMode: ProcessingMode,
+        category: String,
+        // TODO Engine setup shouldn't be inside parameters. It should be rather picked on the stage when user decide where to deploy scenario
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/api/ScenarioParametersEndpoints.scala:66]: Because we already know that it isn't a good thing. To make it as it should be done, we have to:
- Remove the concept of processing type - we should have only components and deployments
- Add an option of choosing engine setup somewhere around the deploy button
- Make components pallet smarter

We decided to not do all these things and focus on moving forward batch processing mode and related things.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/ScenarioParametersService.scala [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+import pl.touk.nussknacker.engine.api.component.ProcessingMode
+import pl.touk.nussknacker.engine.api.process.ProcessingType
+import pl.touk.nussknacker.engine.deployment.EngineSetupName
+import pl.touk.nussknacker.engine.util.Implicits.RichTupleList
+import pl.touk.nussknacker.restmodel.scenariodetails.{ScenarioParameters, ScenarioParametersWithEngineSetupErrors}
+import pl.touk.nussknacker.security.Permission
+import pl.touk.nussknacker.ui.security.api.LoggedUser
+import pl.touk.nussknacker.ui.{NotFoundError, NuDesignerError, UnauthorizedError}
+
+class ScenarioParametersService private (
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/ScenarioParametersService.scala:14]: I don't understand the question :) It is a kind of domain service - it calculates `ScenarioParameters`. What do you mean by "other values"?
- `categoryUnsafe` - Category is the one of these parameters
- `getProcessingTypeWithWritePermission` - This class is also responsible for transition between processing type and scenario parameters in two directions - processingtype -> parameters in factory ('apply') method, opposite direction in this method.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+  "The endpoint for getting scenario parameters combination when" - {
+    "authenticated with write access to all categories should" - {
+      "return scenario parameters combination for all categories" in {
+        given()
+          .basicAuth("admin", "admin")
+          .when()
+          .get(s"$nuDesignerHttpAddress/api/scenarioParametersCombinations")
+          .Then()
+          .statusCode(200)
+          .equalsJsonBody(s"""
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala:32]: What exactly do you mean? Should I remove on empty blank line in the body? Or you would like to add a new line before the string? The second option is impossible because scalafmt corrects this.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+import org.scalatest.freespec.AnyFreeSpecLike
+import pl.touk.nussknacker.test.{
+  NuRestAssureExtensions,
+  NuRestAssureMatchers,
+  PatientScalaFutures,
+  RestAssuredVerboseLogging
+}
+import pl.touk.nussknacker.ui.api.helpers.{NuItTest, NuTestScenarioManager}
+
+class ScenarioParametersApiSpec
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala:14]: 1. What do you mean by missing tests for authorization? There are tests for: access to all categories, access to one category, access to none category.
2. Regarding engine errors:
- It will be hard to do this at this level because it would require showing different combinations of application setup and currently we don't have option to change this in a given part
- But it will be easy to do at the domain service layer and I have a plan to add them in the next PRs. Currently, engine errors are always an empty list because of FIXME in `ProcessingTypeData:124`

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/integration/BaseFlowTest.scala [hunk 5]
```diff
@@ -300,15 +294,16 @@ class BaseFlowTest
...
+    validationResponse.code shouldEqual StatusCode.Ok
+    validationResponse.body should include("Configured property environment (Environment) is missing")
+    validationResponse.body should include("This field value has to be an integer number")
+    validationResponse.body should include("Unknown property unknown")
+    validationResponse.body should include("Property numberOfThreads (Number of threads) has invalid value") //
   }
 
   test("be able to work with fragment with custom class inputs") {
-    val processId = UUID.randomUUID().toString
+    val processId = ProcessName(UUID.randomUUID().toString)
```
[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/integration/BaseFlowTest.scala:305]: I don't want to change every process -> scenario in this file in this PR - doing this one by one doesn't scale. IMO we should do this using some replace all in all files. I already did a lot similar cleanups. Take a look at https://github.com/TouK/nussknacker/pull/5434 . In the new files I'm trying to use scenario everywhere

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala]: Why is it impossible? You can check: pl.touk.nussknacker.ui.api.AppApiSpec or pl.touk.nussknacker.ui.api.ScenarioActivityApiSpec for examples

[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala]: I don't know why :) I add a new line, do reformat and it back to the previous state. Check by yourself and tell me how to deal with this :)

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala:14]: 1. Ah, I mean authentication. You can write test without basic authentication.

[the remark this answers is not in the document]
Comment (arkadius) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala:14]: Ah, you are right. I added it.

File: designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply designer/server/src/test/scala/pl/touk/nussknacker/ui/api/ScenarioParametersApiSpec.scala]: I formatted like this:
```
          .equalsJsonBody(
            s"""
               |[
               |    {
               |        "processingMode": "Unbounded-Stream",
               |        "category": "Category1",
               |        "engineSetup": {
               |            "name": "Flink",
               |            "errors": [
               |
               |            ]
               |        }
               |    },
               |    {
               |        "processingMode": "Unbounded-Stream",
               |        "category": "Category2",
               |        "engineSetup": {
               |            "name": "Flink",
               |            "errors": [
               |
               |            ]
               |        }
               |    }
               |]""".stripMargin
          )
```
and then I commited changes. Scalafmt does not change anything.

File: designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/ScenarioParametersService.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (lukasz-bigorajski) [reply designer/server/src/main/scala/pl/touk/nussknacker/ui/process/processingtype/ScenarioParametersService.scala:14]: I understand the concept behind this class. But if you ask me, what a service is and where it should be located, I would say it is the same as this class except that it can have dependency to external service/repositories etc. and is located in package with other services. I would expect that a class which is located in domain package has name like `ScenarioParameterCalculator` or just `ScenarioParameters`.
Eventually I don't insist on any change, maybe it's only my personal view on this topic.

**Review by lukasz-bigorajski — APPROVED** [review]
